File: 09-Prototype/src/go/circle.go
package main

import "fmt"

type Circle struct {
	baseShape
	Radius int
}

func NewCircle(x, y int, color string, radius int) *Circle {
	return &Circle{
		baseShape: baseShape{X: x, Y: y, Color: color, Tags: []string{}},
		Radius:    radius,
	}
}

func (c *Circle) Clone() Shape {
	return &Circle{
		baseShape: baseShape{
			X:     c.X,
			Y:     c.Y,
			Color:  c.Color,
			Tags:   deepCopyTags(c.Tags), // deep copy the slice
		},
		Radius: c.Radius,
	}
}

func (c *Circle) String() string {
	return fmt.Sprintf("Circle [radius=%d (%d,%d) color=%s tags=%v]",
		c.Radius, c.X, c.Y, c.Color, c.Tags)
}

File: 09-Prototype/src/java/chapter09/Circle.java
package chapter09;

public class Circle extends Shape {
    private int radius;

    public Circle(int x, int y, String color, int radius) {
        super(x, y, color);
        this.radius = radius;
    }

    // Copy constructor
    public Circle(Circle source) {
        super(source);                 // copies x, y, color, tags (deep)
        this.radius = source.radius;   // primitive — copied by value
    }

    @Override
    public Circle clone() {
        return new Circle(this);
    }

    public int getRadius() {
        return radius;
    }

    @Override
    public String toString() {
        return "Circle [radius=" + radius + " " + super.toString() + "]";
    }
}
